pygmt/clib/session.py

```diff
@@ -1359,6 +1360,62 @@ def virtualfile_from_grid(self, grid):
         with self.open_virtual_file(*args) as vfile:
             yield vfile
 
+    def virtualfile_from_data(self, data, x=None, y=None, z=None, check_kind=None):
```

Comment:
A quick question: does it work for modules that requires more than three columns? For example, in `Figure.plot()`, we need to pass `extra_arrays` to GMT API.

Reply:
Excellent point. Perhaps put the `check_kind` parameter in front, so that it becomes `virtualfile_from_data(self, check_kind, data, x, y, z, *extra_arrays)`?

Reply:
I'm thinking if having two functions can make it easier:

- virtualfile_from_table_data: `data` can be `file` or `matrix`, or `x, y, z, extra_arrays` can be `vector`.
- virtualfile_from_grid_data: `data` can be `file`, or `xr.DataArray`

Reply:
I'm OK with one universal function, but you may want to look at some more complicated modules (e.g., `grdtrack` and `plot`). In these modules, it seems the data `kind` is needed outside the `virtualfile_from_data` function.

Reply:
Yeah, `grdtrack` needs a good refactor, and so does `plot`. I think we can refactor this `virtualfile_from_data` function when we refactor those complicated modules, so need to make sure the implementation here is ok to be extended to those functions.

Reply:
OK to me.